wasm/datatypes/memory.py

```diff
@@ -47,21 +53,22 @@ def read(self, location: numpy.uint32, size: numpy.uint32) -> bytes:
                     f"> {len(self.data)}"
                 )
 
-        if end_location > len(self.data):
+        if end_location > self._length_cache:
             raise Trap(
                 f"Attempt to read from out of bounds memory location: {end_location} "
                 f"> {len(self.data)}"
             )
-        return self.data[location:end_location]
+        return memoryview(self.data)[location:end_location]
```

Comment:
Hm, it makes me a little queasy that this enables readers to modify the `MemoryInstance`s `.data` with this returned `memoryview`. I tried to find a clean way to make the `memoryview` readonly, but got stuck. [Another person got stuck in the same way](https://lists.gt.net/python/python/1124122#1124122) and resorted to this mess:

```py

pydll = ctypes.cdll.LoadLibrary("python{}{}".format( 
  sys.version_info.major, sys.version_info.minor
)) 

def ro_memoryview_from_bytearray(buffer): 
  assert isinstance(buffer, bytearray) 
  ptr = ctypes.c_char_p(pydll.PyByteArray_AsString( 
    ctypes.py_object(buffer)
  )) 
  mv_id = pydll.PyMemoryView_FromMemory(ptr, len(buffer), 0) 
  return ctypes.cast(mv_id, py_object).value

Reply:
https://github.com/ethereum/py-wasm/issues/115